### ISSUE TITLE
When transaction end, should we inform GTM firstly ?

### ISSUE THREAD

**Issue by DBInsight** [postgres-x2/postgres-x2#239]
2PC ensure all participants commit or rollback, but 2PC does not ensure consistency --- when CN sent commits to DNs , the true commit time of each DN are different, For example, DN1 commit already, but DN2 may commit in progress, If a concurrent query may see DN1's write, but can not see DN2's. So, 2PC does not ensure global consistency.

For this reason, we have GTM, the snapshot is taken from GTM, So, in above scenario, even if DN1 commits, but CN and GTM, in GTM, we still maintain the transaction is in progress, so we can not see DN's commits, this is theoretical founddation of global consistency.

Unfortunately, PG has SnapShotNow and Same logic as SnapshotNow,for example: HeapTupleSatisfiesUpdate, although the new version of PG has removed SnapshotNow, the HeapTupleStatisfiesUpdate still  exists and impossible to remove, HeapTupleStatisfiesUpdate must see the newest version of tuple, when DN's local transaction commits, heap update will see the newest tuple, because it checks TransactionIdIsInProgress which only checks DN's local proc array, it ignores the GTM's Snapshot completely,  in one DN, it will see xid is committed and concurrent update continues. the SnapshotNow-like logical breaks the  theoretical founddation of global consistency: snapshot must be taken from GTM and tuple visibility rules must honor this snapshot. But SnapshotNow-like logical ignores the global snapshot, so consistency is broken.

the following example can reproduce:
create table test(id int, name varchar);
insert into test values(1, 'aa');
update test set name='bb' where id=1; --txn1, gdb break after DN commits and before inform GTM;

another session;
update test set name='cc' where id=1; --txn2
select \* from test;
will see two lows:
1, 'aa'                              --txn1 is not commit, so xmax is invalid, we will see this row;
1, 'cc'                              --txn2 is commit, xmin is commited, will see this row;

I think, we should inform GTM firstly when transaction end, the global transaction manager should know the transaction's status change firstly. If we find the transaction ends on GTM and inprogress on CN/DN, we should wait CN/DN's local inprogress transaction ends and it will be definitely be.

**Comment by koichi-szk** [comment]
Thank you for the analysis.   We need to modify HeapTupleStatisfiesUpdate to see the snapshot from GTM.    Order of reporting GTM and do real commit operation needs some more thought.  If coordinator fails after the commit operation and before it reports to GTM, GTM detects its failure and remove its GXID from the list.   Afterwords, its GXID does not appear in the snapshot, which is correct behavior.

If the backend reports GTM first, then GTM removes its GXID from snapshot and such snapshot can be used by other backends before actual commit operation has been done.   I don't think it's a correct order.

Any more thoughts?

**Comment by DBInsight** [comment]
not just HeapTupleStatisfiesUpdate, all scenario that call `TransactionIdIsInProgress' will have problem, because`TransactionIdIsInProgress' does not honor global snapshot;

If coordinator fails after the commit operation and before it reports to GTM, GTM detects its failure and remove its GXID from the list. Afterwords, its GXID does not appear in the snapshot, which is correct behavior.
-----it is not true for RXACT_PART_COMMITTED, part commit will record a prepared transaction on GTM, even if GTM detects failure, it does not remove Prepared transaction ; in this case, the example i posted above will unconsistency.

If the backend reports GTM first, then GTM removes its GXID from snapshot and such snapshot can be used by other backends before actual commit operation has been done. I don't think it's a correct order.
---When we get snapshot from GTM, CN/DN should add a check, if there're prepared transactions that has not actual commit, we should wait these transaction ended or invoke pgxc_clean to clear.

**Comment by DBInsight** [comment]
We need to modify HeapTupleStatisfiesUpdate to see the snapshot from GTM. Order of reporting GTM and do real commit operation needs some more thought. 

---It would be best not to modify the original logic of postgres;

thoughts?

**Comment by koichi-szk** [comment]
Postgres does not cover atomic visibility in the case of multi-node transaction.    This is what GTM is doing and minimum modification will be needed to implement it.   I agree not to do any more modification.

**Comment by dierbua** [comment]
maybe,we slove this bug.

**Comment by koichi-szk** [comment]
I believe we should do local commits in coordinator/datanode before reporting to GTM.   If report is done to GTM first, then transaction could be visible before it is really committed.   This can cause other problems especially for updated rows.

**Comment by DBInsight** [comment]
if we inform GTM, if a transaction commit in GTM but in progress in CN/DN, we should wait this transaction end, 

if a transaction in CN/DN not in GTM snapshot, we should wait this transaction end.

**Comment by koichi-szk** [comment]
Yes, my fix is trying to do this.   I'm reviewing the original code if this works with internal locks related to heap update.

In fact, transaction reports to GTM before it "PREPARES", and then report to GTM after the commit.   Issue here is, as you pointed out, conflicting update should wait until the transaction commit is reported to GTM.

**Comment by koichi-szk** [comment]
This is related to the issue #256.
